### PR TITLE
[Ability] Add form change support for Flower Gift

### PR DESCRIPTION
**Pull request by torranx**
## What are the changes the user will see?
Cherrim should now transform during Harsh Sunlight similar to Castform

## Why am I making these changes?
Cherrim does not transform on Harsh Sunlight

## What are the changes from a developer perspective?
Added form change conditions - attached to what Forecast is using

## How to test the changes?
Use Cherrim and set weather to Harsh Sunlight/Sunny

## Checklist
- [x] **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I considered writing automated tests for the issue?
- [ ] If I have text, did I make it translatable and add a key in the English locale file(s)?
- [x] Have I tested the changes (manually)?
    - [x] Are all unit tests still passing? (`npm run test`)
- [ ] Are the changes visual?
  - [ ] Have I provided screenshots/videos of the changes?

### CODE REVIEW

File: src/data/ability.ts [hunk 1]
```diff
@@ -2410,7 +2410,10 @@ export class PostSummonFormChangeByWeatherAbAttr extends PostSummonAbAttr {
    * @returns whether the form change was triggered
    */
   applyPostSummon(pokemon: Pokemon, passive: boolean, simulated: boolean, args: any[]): boolean {
-    if (pokemon.species.speciesId === Species.CASTFORM && this.ability === Abilities.FORECAST) {
+    const isCastformWithForecast = pokemon.species.speciesId === Species.CASTFORM && this.ability === Abilities.FORECAST;
```
Comment (Tempo-anon) [inline src/data/ability.ts:2413]: nit: here and below, I would prefer to have the right half of the statement wrapped in parentheses.
`    const isCastformWithForecast = (pokemon.species.speciesId === Species.CASTFORM && this.ability === Abilities.FORECAST);`

File: src/data/ability.ts [hunk 2]
```diff
@@ -4704,6 +4711,7 @@ function setAbilityRevealed(pokemon: Pokemon): void {
 function getPokemonWithWeatherBasedForms(scene: BattleScene) {
   return scene.getField(true).filter(p =>
     p.hasAbility(Abilities.FORECAST) && p.species.speciesId === Species.CASTFORM
+    || p.hasAbility(Abilities.FLOWER_GIFT) && p.species.speciesId === Species.CHERRIM
```
Comment (Tempo-anon) [inline src/data/ability.ts:4714]: nit: also with the boolean stuff I'd prefer having parentheses to make it clearer

```ts
(p.hasAbility(Abilities.FORECAST) && p.species.speciesId === Species.CASTFORM)
    || (p.hasAbility(Abilities.FLOWER_GIFT) && p.species.speciesId === Species.CHERRIM)
```

**Review by Tempo-anon — COMMENTED** [review]
Wasn't around for when the Forecast PR got merged in but I assume it also works with all the misc edge cases such as weather being suppressed and then no longer being suppressed? If so then this looks good to me, just some personal style nits

**Comment by torranx** [comment]
> Wasn't around for when the Forecast PR got merged in but I assume it also works with all the misc edge cases such as weather being suppressed and then no longer being suppressed? If so then this looks good to me, just some personal style nits

Yes, pesky Air Lock/Cloud Nine - but one edge case is noted on the Forecast PR description (which does have already at least 2 PRs trying to fix it rn)

TLDR: if enemy with air lock/cloud nine didnt receive any attacks and fainted, post faint ability attrs do not proc - hence form change not happening, this is bypassed by PreSwitchOut ab attrs for the player & enemy trainer tho

**Review by Tempo-anon — APPROVED** [review]

**Comment by DayKev** [comment]
Well, if Cherubi evolves into Cherrim, then it won't form-change if the weather is currently sunny when it evolves. Dunno if we care about this edge case.

**Review by DayKev — APPROVED** [review]

**Comment by torranx** [comment]
> Well, if Cherubi evolves into Cherrim, then it won't form-change if the weather is currently sunny when it evolves. Dunno if we care about this edge case.

Yeah this is an unaccounted edge case but I'll leave this as is for now since we currently don't have a system for post evolution.
